### ISSUE TITLE
Register site on proxy service before redirecting to OAuth

### ISSUE THREAD

**Issue by felixarntz** [google/site-kit-wp#2044]
As part of the new user input flow, the onboarding on the proxy service will slightly change. Most importantly, the initial "Welcome" screen will no longer be present. The form on that screen normally registers or looks up a new site if that site is not (already) in possession of site ID and site secret. Per the removal of that step, the plugin must now take care of this registration call to the proxy service and afterwards redirect to the OAuth endpoint as part of the setup flow on its own.

**This is closely related to #4383 and should preferably be worked on in the same sprint.**

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

In the new version of the setup flow (`serviceSetupV2`), there is no longer a "Welcome" screen present on the authentication service when a user gets there for the Site Kit plugin setup; instead they land right away on the OAuth consent screen (before reaching the UI with the 3 steps as usual). The "Welcome" screen used to call another API route to register or look up the site on the service, so since that is no longer going to happen, this call has to be issued from the plugin going forward. A proof-of-concept implementation for parts of this and #2045 already exists [here](https://github.com/google/site-kit-wp/compare/poc/service-setup-v2-migration), so that branch will be referenced in a few places in the ACs below.

**All changes below should only apply if the `serviceSetupV2` feature flag is enabled.**

* The `Google_Proxy::OAUTH2_SITE_URI` endpoint on the service should going forward be used to register or look up the site whenever the plugin setup is started (i.e. on `Google_Proxy::ACTION_SETUP`).
    * Instead of using this route to only sync site fields if they are out of sync, this route should now be called _in any case_.
        * If the site is not connected, the route needs to be called to register or look up the site based on the site fields. From the plugin site there isn't much of a difference for whether the site gets registered (i.e. it has never used the service before) or just looked up (i.e. it has used the service before, but then e.g. was reset so that it doesn't have the site credentials anymore).
        * If the site is already connected, the route needs to be called with the site credentials, in which case it will have pretty much the same purpose as in the existing usage, to update any site fields out of sync.
        * The reason that this call now needs to _always_ happen, even in the "already connected" case above, is that the route will now return the proper URI for where to actually redirect the user to for the setup. It has always done that for the service, but now we need to use the same URI in the plugin. In other words, the new flow is to call that URI, retrieve the redirect URI returned, and then redirect the user to that URI (i.e. no longer use the `Google_Proxy::setup_url()` here).
    * Below some more technical details on the exact requirements:
        * When calling the `Google_Proxy::OAUTH2_SITE_URI` endpoint, it should now _always_ include a few additional body parameters, since it is also used to compose the full URI to redirect the user to. Again, it has always been like that on the service side, but now we need to do that from the plugin side too since it's now used for the same purpose here:
            * In addition to all the fields from `Google_Proxy::get_site_fields()`, the endpoint should now also receive all the fields from `Google_Proxy::get_user_fields()`, the `scope` query parameter (as usual, the space-separated list of required scopes), and a new set of "metadata" fields, most of which so far were only used for the `Google_Proxy::setup_url()`.
            * For which fields exactly are those metadata fields, please look at the `Google_Proxy::get_metadata_fields()` method implemented in the [POC](https://github.com/google/site-kit-wp/compare/poc/service-setup-v2-migration). The list there should be what the implementation here uses too (and potentially it's even worth introducing the exact same method). Brief note on the `serviceVersion` parameter: That's the only one which is entirely new, and it should be passed to tell the service to use its v2 version of the setup endpoint (which will have related behavioral changes and updated UI).
            * It may also be useful to look at the `Google_Proxy::register_site()` method in the [POC](https://github.com/google/site-kit-wp/compare/poc/service-setup-v2-migration) as overall inspiration.
        * The redirect URI from the `Google_Proxy::OAUTH2_SITE_URI` endpoint is not returned in the response body, but in a response header `Redirect-To` (formerly `AMP-Redirect-To`, but let's not use this one here).
            * This is solved in a quite hacky way in the `Google_Proxy::request()` method in the [POC](https://github.com/google/site-kit-wp/compare/poc/service-setup-v2-migration), but this may still be useful as a code example.
        * One additional small but crucial tweak required here (that part is currently **not** implemented in the POC) is to migrate the changes made in #4342 to work with the new approach here. In other words, the `mode=analytics-step` parameter needs to be conditionally present in the data sent to the `Google_Proxy::OAUTH2_SITE_URI` as well. Since `mode` is one of the "metadata" parameters outlined above, it probably makes sense to simply make that particular value filterable, via a new WordPress filter. That filter can then be applied in the `Analytics` class to conditionally alter it, relying on the same logic that was already implemented in #4342. Since this is only relevant for V2 anyway, the `googlesitekit_proxy_setup_url_params` filter usage is no longer relevant - and the `googlesitekit_proxy_setup_url_params` should be removed completely, as it is no longer used anymore anywhere and has caused some confusing code in the past.
    * For the overall implementation of the new redirect behavior, it may help looking at the `Authentication::redirect_to_setup_flow()` method in the [POC](https://github.com/google/site-kit-wp/compare/poc/service-setup-v2-migration).

#### includes/Core/Authentication/Google_Proxy.php
* [x] Add a new constant `HEADER_REDIRECT_TO` that equals to `Redirect-To`.
* [x] Add a new private property `$required_scopes`.
* [x] Add a new public method `with_scopes( array $scopes )` that assigns the incoming scopes to the `$required_scopes` property.
* [x] Update the `request` method to return the response object itself it the `$args['return']` property equals `response`. Return the response object after checking the response status code and before checking whether the body is null-ish or not: https://github.com/google/site-kit-wp/blob/a608bcba230cf33c225b33730b15204aecfad5f6/includes/Core/Authentication/Google_Proxy.php#L252
* [x] Add a new public method `get_metadata_fields()` that:
   * composes a new array with the following properties:
      * `supports` - space-separated list of supported features returned by the `get_supports()` method.
      * `nonce` - a nonce created for the `ACTION_SETUP` action.
      * `mode` - an empty string.
      * `hl` - the current user locale that can be received using `$this->context->get_locale( 'user' )` call.
      * `application_name` - the application name that can be obtained by calling the `get_application_name` method.
      * `service_version` - a string that indicates the service version, it should be `v2` if the `serviceSetupV2` feature flag is enabled.
   * applies the `googlesitekit_proxy_setup_mode` filter to the `mode` property of the metadata array.
   * Returns the metadata array.
* [x] Add a new public method `register_site()` that:
   * sends a request to the `Google_Proxy::OAUTH2_SITE_URI` endpoint with the following arguments:
      * `return` should be `response`.
      * `body` should be a combined array that includes properties from arrays returned by `get_site_fields()`, `get_user_fields()`, and `get_metadata_fields()` methods, plus the `scopes` property that containes the whitespace separated string of `$required_scopes` passed via the `with_scopes` method.
   * takes the response object and returns it as is if it is an error
   * otherwise retrieves the `Redirect-To` header using the `wp_remote_retrieve_header` function and returns it.
* [x] Update the `sync_site_fields` method to do the following:
   * Add `'return' => 'response'` property to the request arguments array.
   * Update the `body` property of the arguments list to compose the same array as defined in the `register_site` method.
   * Assign the result of the `request` method call to a variable and return it as is if it is a WP error.
   * Otherwise retrieve the `Redirect-To` header using the `wp_remote_retrieve_header` function and return it.

#### includes/Core/Authentication/Setup_V2.php
* [x] Update the `handle_action_setup_start` method to do the following:
    * Get the required scopes array using `$this->authentication->get_oauth_client()->get_required_scopes()` call and add it to the proxy class calling `$this->google_proxy->with_scopes( ... )` method.
    * Assign the returned value of the `$this->google_proxy->sync_site_fields( ... )` function call to a new variable `$oauth_setup_redirect`.
    * Update the `if ( $this->credentials->has() )` condition to have an `else` block in which call the `$this->google_proxy->register_site()` function assigning the returned value to the `$oauth_setup_redirect` variable.
    * Call the `wp_die( 'The request to the authentication proxy has failed. Please, try again later.' )` function if the `$oauth_setup_redirect` is an error or not an URL.
    * Otherwise, redirect to the `$oauth_setup_redirect` URL instead of calling the `$this->redirect_to_proxy()` method.

#### includes/Modules/Analytics.php
* [x] Add a new hook for the `googlesitekit_proxy_setup_mode` filter that returns `analytics-step` if the `serviceSetupV2` feature flag is activated and the Analytics module is not connected. Otherwise return the original value.

## Test Coverage

* Update existing tests affected by these changes.

## QA Brief

* Enable the `serviceSetupV2` feature flag.
* Use the staging version of the proxy.
* Test that the following processes work correctly:
    * Going through setup for a completely new site.
    * Reconnecting a previously set up site after a reset.

## Changelog entry

* Register site on the proxy before redirecting to it, in support for V2 setup flow.

**Comment by aaemnnosttv** [comment]
> Per the removal of that step, the plugin must not take care of this registration call to the proxy service and afterwards redirect to the OAuth endpoint as part of the setup flow on its own

@felixarntz I think you mean "the plugin must take care of this registration call"?

**Comment by eugene-manuilov** [comment]
@felixarntz, IB is ready for review.

**Comment by felixarntz** [comment]
@eugene-manuilov 

> Add a new constant `HEADER_AMP_REDIRECT_TO` that equals to `AMP-Redirect-To`.

Let's call the constant just `HEADER_REDIRECT_TO`. The fact that it's prefixed with `AMP-` is for proxy-related reasons right now, it's really just the name of the header to look at for where to redirect. I recently discussed this with @aaemnnosttv, and we're soon going to change that actually (see https://app.asana.com/0/1198998509229289/1201440994101245/f), another reason to not include `AMP_` here since this doesn't have to do anything with AMP for the plugin purpose.

> Update the `request` method to return the response object itself it the `$args['return']` property equals `response`.

Can you outline this in more detail? What are possible values for the new `return` arg? Will the default then be `body` (to return the response body which is what happens today), or what should it be?

> `mode` - just an empty

Per the recommendation in the ACs, it may be better to introduce a dedicated filter here to use in the `Analytics` class. The `googlesitekit_proxy_setup_url_params` filter you're referencing here will be removed as part of #4383 and #4376, as it makes the code for composing the setup URL unnecessarily hard to follow. A filter for the `mode` is more specific to this purpose anyway.

A general note on everything in the **includes/Core/Authentication/Authentication.php** section: The code here is significantly going to change in #4383 and #4376, which the issue here depends on, so I don't think it will eventually apply. I suggest we hold off of any IB changes for that part for now and wait until #4383 and #4376 at least have their IBs fully defined.

**Comment by eugene-manuilov** [comment]
> Can you outline this in more detail? What are possible values for the new `return` arg? Will the default then be `body` (to return the response body which is what happens today), or what should it be?

@felixarntz, yes, `body` is the default returning value and it will be returned if the `return` param is not set.

**Comment by eclarke1** [comment]
@felixarntz @FlicHollis I'm going to remove this from this sprint as it is blocked by #4376 which is also still in IB right now. This is to allow space for the Unified Dashboard P0 launch blockers. Let me know if you disagree?

**Comment by felixarntz** [comment]
@eclarke1 Makes sense, let's try to wrap up #4376 in this sprint (cc @techanvil) so that we can work on this one the sprint after.

**Comment by felixarntz** [comment]
@eugene-manuilov IB mostly LGTM, a few comments:

> Add a new constant `HEADER_REDIRECT_TO` that equals to `AMP-Redirect-To`

Per a recent change on the proxy, it will now also provide a simple `Redirect-To` header with the same content, so that we don't need to use the `AMP-Redirect-To` header here, which would be a bit odd. I've updated this in the ACs and your IB already.

> Update the `request` method to return the response object itself it the `$args['return']` property equals `response`

While this works, I think it would be potentially wasteful to first read and JSON-decode the body and then not return it. I understand we have to do that for the condition that tries to return an error from the JSON body, so we _do_ indeed need to keep it like you're saying, but I think it would be nicer if we also returned the `$body` since we already have it. What do you think about the alternative idea of having `$args['return']` support an array? The default would be `array( 'body' )`, but you could also pass  e.g. `array( 'headers', 'body' )` and then the method would return the array of all response headers in addition to the `$body`. In our case we could then use `array( 'headers' )` to only get the headers returned.

> applies the `googlesitekit_proxy_metadata_fields` filter to the metadata array and returns it

I'm not sure we should add this filter for _all_ the metadata fields. We only need one to modify the `mode` parameter, the other ones shouldn't be modifiable at this point. So I think it would be more appropriate and simpler to use a specific filter (e.g. `googlesitekit_proxy_setup_mode`).

> Add a new public method `register_site( array $scopes = array() )` that

Related to this _and_ also to the added parameter for `sync_site_fields`, I'm not a fan of having scopes being passed to the method, which will "permanently" bake this into the method signature. This is how I did it in the POC, but I don't think it's a particularly clean approach. The scopes requested aren't really related to syncing or registering the site (although of course it _is_ technically necessary to pass them), that's why I'd prefer if we could specify them another way. Potentially, it would be better to have another metadata field `scopes`. This field could then either be populated through a new filter (we would for now always pass the same scopes: the currently required scopes, so this could be a single unconditional `add_filter` somewhere), or we could somehow pass the required scopes to `Google_Proxy` or make them otherwise available in that class. Let me know what you think - I'd prefer if we could find a way where the scopes don't have to be provided in the method call, for neither `register_site` nor `sync_site_fields`.

> Add a new hook for the `googlesitekit_proxy_metadata_fields` filter that calls the `update_propxy_setup_mode` method.

See my above note, I think this would be better as a filter for just the "proxy setup mode" (aside: note a tiny typo here).

**Comment by eugene-manuilov** [comment]
> Potentially, it would be better to have another metadata field `scopes`. This field could then either be populated through a new filter (we would for now always pass the same scopes: the currently required scopes, so this could be a single unconditional `add_filter` somewhere), or we could somehow pass the required scopes to `Google_Proxy` or make them otherwise available in that class.

@felixarntz I have added a new method to the `Google_Proxy` class that sets scopes that should be used within `sync_site_fields` and `register_site` methods. Does it sound good to you?

**Comment by felixarntz** [comment]
Thanks @eugene-manuilov, yes that sounds like a solid solution to me. IB ✅ 

(there was just two places in the IB where you had `$require_scopes` as variable name, I fixed that to `$required_scopes`)

**Comment by eugene-manuilov** [comment]
@felixarntz this is ready for code review. Do you mind reviewing it?

**Comment by eugene-manuilov** [comment]
@wpdarren I think this is expected, but @felixarntz can say for sure.

**Comment by wpdarren** [comment]
### QA Update: ⚠️ 

@eugene-manuilov @felixarntz I'd like to run another round of QA on this ticket but in addition to the [observation above](https://github.com/google/site-kit-wp/issues/2044#issuecomment-1028940961) I noticed when connecting Analytics for the first time, at stage 4 of the setup an error message appeared in the console. 

![image](https://user-images.githubusercontent.com/73545194/152348269-64639599-3a91-455c-920d-10d1659dae70.png)

Despite the error, I was able to create a new Analytics account (+ UA and GA4 properties) and when I disconnected the same account was recognised automatically.  Would appreciate your thoughts.

**Comment by felixarntz** [comment]
@wpdarren Hmm that seems unrelated, since it's a failed request coming from the `gtag` snippet on the proxy service I think. Let me check that separately, but we don't need to block this issue on that.

**Comment by wpdarren** [comment]
### QA Update: ⚠️ 

@felixarntz I tested this on two new sites and the console error that I [mentioned here](https://github.com/google/site-kit-wp/issues/2044#issuecomment-1028974656) didn't appear on either of them.  

My only other observation is that when you do not give permission when the oAuth pages are displayed, you are sent back to the dashboard as expected. When I clicked on the link to redo the plugin set up, I'm not taken to the new redesigned set up screen.  The URL is the same though `/wp-admin/admin.php?page=googlesitekit-splash` Thought this was a little odd - maybe it's because we do not reset the site in this scenario. Thought I should mention it. 

Other than this, it's looking good. 

![image](https://user-images.githubusercontent.com/73545194/152508482-99d5ca22-2876-42fe-8ff5-8e952d9994ff.png)

![image](https://user-images.githubusercontent.com/73545194/152508505-65ed830f-9e7a-4f41-8138-84c2f1e08899.png)

**Verified:**

With `serviceSetupV2` feature flag enabled and using the staging version of the proxy.

Successfully went through the setup for a completely new site. I tested this with and without Analytics. 
Successfully went through reconnecting a previously set up site after a reset. I tested this with and without Analytics. 
Checked that when reconnecting that the correct analytics account was selected by default. 
Checked that I was able to change the Analytics account and properties.
Checked that the correct Analytics and Search Console settings were appearing.
Checked the flow when permission was not given and that the user is sent to the dashboard.

**Comment by felixarntz** [comment]
@wpdarren 
> When I clicked on the link to redo the plugin set up, I'm not taken to the new redesigned set up screen.

Can you clarify what you mean? Where does it take you instead? Isn't the screenshot you have above the new redesigned set up screen?

**Comment by wpdarren** [comment]
> Can you clarify what you mean? Where does it take you instead? Isn't the screenshot you have above the new redesigned set up screen?

@felixarntz When cancelling or not giving permission I am redirected to the WP Dashboard. 

![image](https://user-images.githubusercontent.com/73545194/152762621-8df770fe-c30e-422b-b597-ea1cad32e906.png)

When I clicked on the `redo the plugin setup` link, I was expecting to see the screen below. Rather than the screen in my previous comment. Is this because we're using the staging version of the proxy? Wasn't sure so wanted to check. The URL is `/admin.php?page=googlesitekit-splash` that I am landing on. 

Does this make sense?

This is the screen that appears when `serviceSetupV2` is enabled.

<img width="1520" alt="sk" src="https://user-images.githubusercontent.com/73545194/152762120-331f7c9b-050b-43fc-ad1d-29cbc74e87d5.png">

**Comment by felixarntz** [comment]
@wpdarren Ahh, no this is correct. It's because the version with the Analytics checkbox is only shown if Analytics isn't active yet. Since you've decided to activate it before, that happens right away, so it persists even if the rest of the setup fails.

So I think this is all good for now. Although we could evaluate whether we would want to try to align the two versions of the screen a bit more. I think the particularly confusing thing is that the graphic on the right is a different one. cc @marrrmarrr

**Comment by wpdarren** [comment]
QA Update: ✅ 

**Verified:**

With `serviceSetupV2` feature flag enabled and using the staging version of the proxy.

- Successfully went through the setup for a completely new site. I tested this with and without Analytics. 
- Successfully went through reconnecting a previously set up site after a reset. I tested this with and without Analytics. 
- Successfully wet through adding a new user and having them login and connect SK.
- Checked that when reconnecting that the correct analytics account was selected by default. 
- Checked that I was able to change the Analytics account and properties.
- Checked that the correct Analytics and Search Console settings were appearing.
- Checked the flow when permission was not given and that the user is sent to the dashboard.
